File: controllers.py
from interp_logic import Interpolation
import rospy
from threading import Thread
from crkt_common import *
from robot_cmd_ifc import RobotCmdIfc
from robot_state_ifc import RobotStateIfc
import tf_conversions


# Each sub-controller must provide a dict() of all the methods with the method
# name as the key and the method handle as the value so that this Controller
# class and append all the dict() items in a unified way


class ControllerData:
    def __init__(self, controller_name, robot_cmd_ifc, robot_state_ifc):
        self.interpolater = Interpolation()
        self._active = False
        self._controller_name = controller_name
        # New we are connecting the state methods using the naming from the crtk_name of cmd
        # There is no crtk_name for relative 'r' feedback as of yet, so if the command is relative, always bind the
        # position 'p' state method
        # Additionally, there is not 'p', 'v', 'f' or 'r' for joint space feedback, we only have joint states,
        # so if that's the command always bind to 's' for op_space
        op_space_str = get_opspace_as_str(controller_name)
        controller_str = get_controller_as_str(controller_name)
        controller_feedback_str = controller_str

        if op_space_str is 'j':
            controller_feedback_str = 's'
            if controller_str is 'p':
                self._conversion_method = np_array_to_joint_state_pos
            elif controller_str is 'v':
                self._conversion_method = np_array_to_joint_state_vel
            elif controller_str is 'f':
                self._conversion_method = np_array_to_joint_state_effort
            elif controller_str is 'r':
                self._conversion_method = np_array_to_joint_state_pos
                controller_str = 'p'
            else:
                raise Exception('DONTBEANONION')
        elif op_space_str is 'c':
            self._conversion_method = np_array_to_transform_stamped
            if controller_str is 'r':
                controller_str = 'p'
                controller_feedback_str = 'p'

        state_method_name = 'measured_' + op_space_str + controller_feedback_str
        command_method_name = 'servo_' + op_space_str + controller_str
        self.get_robot_state = robot_state_ifc.get_method_by_name(state_method_name)
        self.set_robot_command = robot_cmd_ifc.get_method_by_name(command_method_name)
        self._dt = 1.0
        self._dt_max = 5.0
        self._last_call_time = 0.0

    def set_active(self):
        self._active = True

    def is_active(self):
        return self._active

    def set_idle(self):
        self._active = False

    def set_dt_max(self, dt_max):
        if 1.0 < dt_max < 10.0:
            self._dt_max = dt_max

    def xt(self, t):
        xt = self.interpolater.get_interpolated_x(t)
        return self._conversion_method(xt)

    def calculate_dt(self, cur_time):
        dt = cur_time - self._last_call_time
        self._last_call_time = cur_time
        if 0 < dt <= self._dt_max:
            self._dt = dt
        return self._dt


class Interpolate(object):
    def __init__(self, robot_cmd_ifc, robot_state_ifc, t_start):
        self._t_start = t_start
        self._methods_dict = dict()
        self._methods_dict['interpolate_cp'] = self.interpolate_cp
        self._methods_dict['interpolate_cr'] = self.interpolate_cr
        self._methods_dict['interpolate_cv'] = self.interpolate_cv
        self._methods_dict['interpolate_cf'] = self.interpolate_cf
        self._methods_dict['interpolate_jp'] = self.interpolate_jp
        self._methods_dict['interpolate_jr'] = self.interpolate_jr
        self._methods_dict['interpolate_jv'] = self.interpolate_jv
        self._methods_dict['interpolate_jf'] = self.interpolate_jf

        self._cp_ctrl = ControllerData('interpolate_cp', robot_cmd_ifc, robot_state_ifc)
        self._cr_ctrl = ControllerData('interpolate_cr', robot_cmd_ifc, robot_state_ifc)
        self._cv_ctrl = ControllerData('interpolate_cv', robot_cmd_ifc, robot_state_ifc)
        self._cf_ctrl = ControllerData('interpolate_cf', robot_cmd_ifc, robot_state_ifc)
        self._jp_ctrl = ControllerData('interpolate_jp', robot_cmd_ifc, robot_state_ifc)
        self._jr_ctrl = ControllerData('interpolate_jr', robot_cmd_ifc, robot_state_ifc)
        self._jv_ctrl = ControllerData('interpolate_jv', robot_cmd_ifc, robot_state_ifc)
        self._jf_ctrl = ControllerData('interpolate_jf', robot_cmd_ifc, robot_state_ifc)

        self._controller_data_list = [self._cp_ctrl,
                                      self._cr_ctrl,
                                      self._cv_ctrl,
                                      self._cf_ctrl,
                                      self._jp_ctrl,
                                      self._jr_ctrl,
                                      self._jv_ctrl,
                                      self._jf_ctrl]

        self._thread = Thread(target=self._execute)
        self._thread.start()

    def interpolate_cp(self, cmd):
        state = self._cp_ctrl.get_robot_state()
        if cmd is not None and state is not None:
            p0 = transform_stamped_to_np_array(state)
            pf = transform_stamped_to_np_array(cmd)
            v0 = [0, 0, 0, 0, 0, 0]
            vf = [0, 0, 0, 0, 0, 0]
            a0 = [0, 0, 0, 0, 0, 0]
            af = [0, 0, 0, 0, 0, 0]

            cur_time = rospy.Time.now().to_sec()
            t0 = cur_time - self._t_start
            tf = t0 + self._cp_ctrl.calculate_dt(t0)
            self._cp_ctrl.interpolater.compute_interpolation_params(p0, pf, v0, vf, a0, af, t0, tf)
            self._cp_ctrl.set_active()
        pass

    def interpolate_cr(self, cmd):
        state = self._cr_ctrl.get_robot_state()
        if cmd is not None and state is not None:
            # t_robot = transform_stamped_to_frame(state)
            # t_relative = transform_stamped_to_frame(cmd)
            cmd = transform_stamped_to_frame(state) * transform_stamped_to_frame(cmd)
            p0 = transform_stamped_to_np_array(state)
            pf = frame_to_np_array(cmd)
            v0 = [0, 0, 0, 0, 0, 0]
            vf = [0, 0, 0, 0, 0, 0]
            a0 = [0, 0, 0, 0, 0, 0]
            af = [0, 0, 0, 0, 0, 0]

            cur_time = rospy.Time.now().to_sec()
            t0 = cur_time - self._t_start
            tf = t0 + self._cr_ctrl.calculate_dt(t0)
            self._cr_ctrl.interpolater.compute_interpolation_params(p0, pf, v0, vf, a0, af, t0, tf)
            self._cr_ctrl.set_active()
        pass

    def interpolate_cv(self, cmd):
        pass

    def interpolate_cf(self, cmd):
        state = self._cf_ctrl.get_robot_state()
        if cmd is not None and state is not None:
            f0 = transform_stamped_to_np_array(state)
            ff = transform_stamped_to_np_array(cmd)
            df0 = [0, 0, 0, 0, 0, 0]
            dff = [0, 0, 0, 0, 0, 0]
            ddf0 = [0, 0, 0, 0, 0, 0]
            ddff = [0, 0, 0, 0, 0, 0]

            cur_time = rospy.Time.now().to_sec()
            t0 = cur_time - self._t_start
            tf = t0 + self._cf_ctrl.calculate_dt(t0)
            self._cf_ctrl.interpolater.compute_interpolation_params(f0, ff, df0, dff, ddf0, ddff, t0, tf)
            self._cf_ctrl.set_active()
        pass

    def interpolate_jp(self, cmd):
        state = self._jp_ctrl.get_robot_state()
        if cmd is not None and state is not None:
            j0 = state.position
            jf = cmd.position
            v0 = [0, 0, 0, 0, 0, 0]
            vf = [0, 0, 0, 0, 0, 0]
            a0 = [0, 0, 0, 0, 0, 0]
            af = [0, 0, 0, 0, 0, 0]

            cur_time = rospy.Time.now().to_sec()
            t0 = cur_time - self._t_start
            tf = t0 + self._jp_ctrl.calculate_dt(t0)
            self._jp_ctrl.interpolater.compute_interpolation_params(j0, jf, v0, vf, a0, af, t0, tf)
            self._jp_ctrl.set_active()
        pass

    def interpolate_jr(self, cmd):
        state = self._jr_ctrl.get_robot_state()
        if cmd is not None and state is not None:
            j0 = np.array(state.position)
            jf = j0 + np.array(cmd.position)
            v0 = [0, 0, 0, 0, 0, 0]
            vf = [0, 0, 0, 0, 0, 0]
            a0 = [0, 0, 0, 0, 0, 0]
            af = [0, 0, 0, 0, 0, 0]

            cur_time = rospy.Time.now().to_sec()
            t0 = cur_time - self._t_start
            tf = t0 + self._jr_ctrl.calculate_dt(t0)
            self._jr_ctrl.interpolater.compute_interpolation_params(j0, jf, v0, vf, a0, af, t0, tf)
            self._jr_ctrl.set_active()
        pass

    def interpolate_jv(self, cmd):
        pass

    def interpolate_jf(self, cmd):
        pass

    def _execute(self):
        while not rospy.is_shutdown():
            for controller_data in self._controller_data_list:
                if controller_data.is_active():
                    t = rospy.Time.now().to_sec() - self._t_start
                    while t < controller_data.interpolater.get_tf():
                        data = controller_data.xt(t)
                        controller_data.set_robot_command(data)
                        t = rospy.Time.now().to_sec() - self._t_start
                    controller_data.set_idle()

    def get_methods_dict(self):
        return self._methods_dict


class Servo(object):
    def __init__(self, robot_cmd_ifc, t_start):
        self._methods_dict = dict()
        self._methods_dict['servo_cp'] = self.servo_cp
        self._methods_dict['servo_cr'] = self.servo_cr
        self._methods_dict['servo_cv'] = self.servo_cv
        self._methods_dict['servo_cf'] = self.servo_cf
        self._methods_dict['servo_jp'] = self.servo_jp
        self._methods_dict['servo_jr'] = self.servo_jr
        self._methods_dict['servo_jv'] = self.servo_jv
        self._methods_dict['servo_jf'] = self.servo_jf

    def servo_cp(self):
        pass

    def servo_cr(self):
        pass

    def servo_cv(self):
        pass

    def servo_cf(self):
        pass

    def servo_jp(self):
        pass

    def servo_jr(self):
        pass

    def servo_jv(self):
        pass

    def servo_jf(self):
        pass

    def get_methods_dict(self):
        return self._methods_dict


class Move(object):
    def __init__(self, robot_cmd_ifc, robot_state_ifc, t_start):
        self._t_start = t_start
        self._methods_dict = dict()
        self._methods_dict['move_cp'] = self.move_cp
        self._methods_dict['move_cr'] = self.move_cr
        self._methods_dict['move_jp'] = self.move_jp
        self._methods_dict['move_jr'] = self.move_jr

        self._cp_ctrl = ControllerData('move_cp', robot_cmd_ifc, robot_state_ifc)
        self._cr_ctrl = ControllerData('move_cr', robot_cmd_ifc, robot_state_ifc)
        self._jp_ctrl = ControllerData('move_jp', robot_cmd_ifc, robot_state_ifc)
        self._jr_ctrl = ControllerData('move_jr', robot_cmd_ifc, robot_state_ifc)

        self._controller_data_list = [self._cp_ctrl,
                                      self._cr_ctrl,
                                      self._jp_ctrl,
                                      self._jr_ctrl]

        self._thread = Thread(target=self._execute)
        self._thread.start()

    def move_cp(self, cmd):
        state = self._cp_ctrl.get_robot_state()
        if cmd is not None and state is not None:
            p0 = transform_stamped_to_np_array(state)
            pf = transform_stamped_to_np_array(cmd)
            v0 = [0, 0, 0, 0, 0, 0]
            vf = [0, 0, 0, 0, 0, 0]
            a0 = [0, 0, 0, 0, 0, 0]
            af = [0, 0, 0, 0, 0, 0]

            cur_time = rospy.Time.now().to_sec()
            t0 = cur_time - self._t_start
            tf = t0 + self._cp_ctrl.calculate_dt(t0)
            self._cp_ctrl.interpolater.compute_interpolation_params(p0, pf, v0, vf, a0, af, t0, tf)
            self._cp_ctrl.set_active()
            while self._cp_ctrl.is_active():
                a = 0 # Wait/Block
        pass

    def move_cr(self, cmd):
        state = self._cr_ctrl.get_robot_state()
        if cmd is not None and state is not None:
            cmd = transform_stamped_to_frame(state) * transform_stamped_to_frame(cmd)
            p0 = transform_stamped_to_np_array(state)
            pf = frame_to_np_array(cmd)
            v0 = [0, 0, 0, 0, 0, 0]
            vf = [0, 0, 0, 0, 0, 0]
            a0 = [0, 0, 0, 0, 0, 0]
            af = [0, 0, 0, 0, 0, 0]

            cur_time = rospy.Time.now().to_sec()
            t0 = cur_time - self._t_start
            tf = t0 + self._cr_ctrl.calculate_dt(t0)
            self._cr_ctrl.interpolater.compute_interpolation_params(p0, pf, v0, vf, a0, af, t0, tf)
            self._cr_ctrl.set_active()
            while self._cr_ctrl.is_active():
                a = 0 # Wait/Block
        pass

    def move_jp(self, cmd):
        state = self._jp_ctrl.get_robot_state()
        if cmd is not None and state is not None:
            j0 = state.position
            jf = cmd.position
            v0 = [0, 0, 0, 0, 0, 0]
            vf = [0, 0, 0, 0, 0, 0]
            a0 = [0, 0, 0, 0, 0, 0]
            af = [0, 0, 0, 0, 0, 0]

            cur_time = rospy.Time.now().to_sec()
            t0 = cur_time - self._t_start
            tf = t0 + self._jp_ctrl.calculate_dt(t0)
            self._jp_ctrl.interpolater.compute_interpolation_params(j0, jf, v0, vf, a0, af, t0, tf)
            self._jp_ctrl.set_active()
            while self._jp_ctrl.is_active():
                a = 0 # Wait/Block
        pass

    def move_jr(self, cmd):
        state = self._jr_ctrl.get_robot_state()
        if cmd is not None and state is not None:
            j0 = np.array(state.position)
            jf = j0 + np.array(cmd.position)
            v0 = [0, 0, 0, 0, 0, 0]
            vf = [0, 0, 0, 0, 0, 0]
            a0 = [0, 0, 0, 0, 0, 0]
            af = [0, 0, 0, 0, 0, 0]

            cur_time = rospy.Time.now().to_sec()
            t0 = cur_time - self._t_start
            tf = t0 + self._jr_ctrl.calculate_dt(t0)
            self._jr_ctrl.interpolater.compute_interpolation_params(j0, jf, v0, vf, a0, af, t0, tf)
            self._jr_ctrl.set_active()
            while self._jr_ctrl.is_active():
                a = 0 # Wait/Block
        pass

    def _execute(self):
        while not rospy.is_shutdown():
            for controller_data in self._controller_data_list:
                if controller_data.is_active():
                    t = rospy.Time.now().to_sec() - self._t_start
                    while t < controller_data.interpolater.get_tf():
                        data = controller_data.xt(t)
                        controller_data.set_robot_command(data)
                        t = rospy.Time.now().to_sec() - self._t_start
                    controller_data.set_idle()

    def get_methods_dict(self):
        return self._methods_dict


class Controllers(object):
    def __init__(self, namespace, arm_name, t_start=0.0):
        self._robot_cmd_ifc = RobotCmdIfc(namespace, arm_name)
        self._robot_state_ifc = RobotStateIfc(namespace, arm_name)
        self.controllers_list = [Interpolate(self._robot_cmd_ifc, self._robot_state_ifc, t_start),
                                 Servo(self._robot_cmd_ifc, t_start),
                                 Move(self._robot_cmd_ifc, self._robot_state_ifc, t_start)]
        self._methods_dict = dict()
        for controller in self.controllers_list:
            self._methods_dict.update(controller.get_methods_dict())
        pass

    def get_method_by_name(self, method_name):
        return self._methods_dict[method_name]

    def disconnect(self):
        self._robot_state_ifc.clean()
        self._robot_cmd_ifc.clean()

File: src/Controllers.cpp
#include "motion_ifc/Controllers.h"

///
/// \brief SubControllersCommon::_initialized
///
bool ControllerManager::_initialized = false;

///
/// \brief SubControllersCommon::SubControllersCommon
///
ControllerManager::ControllerManager():
    rate(1000),
    RobotCmd(crtk::get_name_space(), crtk::get_arm_name() ),
    RobotState(crtk::get_name_space(), crtk::get_arm_name() ){
    if (!_initialized){
        _initialized = true;
        bind_robot_io("robot_cp", cpCtrl);
        bind_robot_io("robot_cr", crCtrl);
        bind_robot_io("robot_cv", cvCtrl);
        bind_robot_io("robot_cf", cfCtrl);

        bind_robot_io("robot_jp", jpCtrl);
        bind_robot_io("robot_jr", jrCtrl);
        bind_robot_io("robot_jv", jvCtrl);
        bind_robot_io("robot_jf", jfCtrl);

        execTh = boost::thread(boost::bind(&Interpolate::execute, this));
    }
}

///
/// \brief SubControllersCommon::bind_robot_io
/// \param interface_name
/// \param ctrlrBase
///
void ControllerManager::bind_robot_io(string interface_name, CtrlrBasePtr &ctrlrBase){
    std::vector<std::string> crtk_str = split_str(interface_name, '_');
    char op_space = crtk_str[1][0];
    char controller = crtk_str[1][1];

    char cmd_controller = controller;
    char state_controller = controller;

    if (op_space == 'c'){
        switch (controller){
        case 'p':
            ctrlrBase = CtrlrBasePtr(new ControllerData<_cp_data_type,  _cp_data_type>(interface_name));
            break;
        case 'r':
            ctrlrBase = CtrlrBasePtr(new ControllerData<_cr_data_type,  _cr_data_type>(interface_name));
            cmd_controller = 'p';
            state_controller = 'p';
            break;
        case 'v':
            ctrlrBase = CtrlrBasePtr(new ControllerData<_cv_data_type,  _cv_data_type>(interface_name));
            break;
        case 'f':
            ctrlrBase = CtrlrBasePtr(new ControllerData<_cf_data_type,  _cf_data_type>(interface_name));
            break;
        default:
            throw "The specified format isnt understood";
        }
    }
    else if (op_space == 'j'){
        state_controller = 's';
        switch (controller){
        case 'p':
            ctrlrBase = CtrlrBasePtr(new ControllerData<_jp_data_type,  _js_data_type>(interface_name));
            break;
        case 's':
            ctrlrBase = CtrlrBasePtr(new ControllerData<_jp_data_type,  _js_data_type>(interface_name));
            break;
        case 'r':
            ctrlrBase = CtrlrBasePtr(new ControllerData<_jr_data_type,  _js_data_type>(interface_name));
            cmd_controller = 'p';
            break;
        case 'v':
            ctrlrBase = CtrlrBasePtr(new ControllerData<_jv_data_type,  _js_data_type>(interface_name));
            break;
        case 'f':
            ctrlrBase = CtrlrBasePtr(new ControllerData<_jf_data_type,  _js_data_type>(interface_name));
            break;
        default:

            throw "The specified format isnt understood";
        }

    }
    std::string robot_state_search_str = "measured_";
    robot_state_search_str += op_space;
    robot_state_search_str += state_controller;
    std::string robot_cmd_search_str = "servo_";
    robot_cmd_search_str += op_space;
    robot_cmd_search_str += cmd_controller;

    if(DEBUG){
        std::cout << "Robot CMD Method: " << robot_cmd_search_str << std::endl;
        std::cout << "Robot STATE Method: " << robot_state_search_str << std::endl;
    }

    ctrlrBase->robot_cmd_method = RobotCmd::get_method_by_name(robot_cmd_search_str);
    ctrlrBase->robot_state_method = RobotState::get_method_by_name(robot_state_search_str);
    ctrlrBase->robot_cmd_ifc = RobotCmd::get_interface_by_name(robot_cmd_search_str);
    ctrlrBase->robot_state_ifc = RobotState::get_interface_by_name(robot_state_search_str);

    vec_ctrlrs.push_back(ctrlrBase);

}

///
/// \brief SubControllersCommon::execute
///
void ControllerManager::execute(){
    while (ros::ok()){
        for (std::vector<CtrlrBasePtr>::iterator it = vec_ctrlrs.begin() ; it != vec_ctrlrs.end() ; it++){
            if ((*it)->is_active()){
                (*it)->set_idle();
                double t = ros::Time::now().toSec();
                while ( (*it)->interpolater.get_t0() <= t && t <= (*it)->interpolater.get_tf()){
                    t = ros::Time::now().toSec();
                    (*it)->cmd_robot(t);
                    rate.sleep();
                }
            }
        }
    }
}

///
/// \brief Interpolate::Interpolate
///
Interpolate::Interpolate(){
    method_map["interpolate_cp"] = FcnHandleBasePtr( new FcnHandle<_cp_data_type>(&Interpolate::interpolate_cp, this));
    method_map["interpolate_cr"] = FcnHandleBasePtr( new FcnHandle<_cr_data_type>(&Interpolate::interpolate_cr, this));
    method_map["interpolate_cv"] = FcnHandleBasePtr( new FcnHandle<_cv_data_type>(&Interpolate::interpolate_cv, this));
    method_map["interpolate_cf"] = FcnHandleBasePtr( new FcnHandle<_cf_data_type>(&Interpolate::interpolate_cf, this));

    method_map["interpolate_jp"] = FcnHandleBasePtr( new FcnHandle<_jp_data_type>(&Interpolate::interpolate_jp, this));
    method_map["interpolate_jr"] = FcnHandleBasePtr( new FcnHandle<_jr_data_type>(&Interpolate::interpolate_jr, this));
    method_map["interpolate_jv"] = FcnHandleBasePtr( new FcnHandle<_jv_data_type>(&Interpolate::interpolate_jv, this));
    method_map["interpolate_jf"] = FcnHandleBasePtr( new FcnHandle<_jf_data_type>(&Interpolate::interpolate_jf, this));
}

///
/// \brief Interpolate::interpolate_cp
/// \param data
///
void Interpolate::interpolate_cp(_cp_data_type &data){
    if(DEBUG) std::cout << "Called: " << __FUNCTION__ << std::endl;
    _cp_data_type robot_state;
    cpCtrl->robot_state_ifc->get_data(robot_state);
    double t0 = ros::Time::now().toSec();
    double tf = t0 + cpCtrl->compute_dt(t0);
    StateSpace ss0 = *cpCtrl->serialize(robot_state);
    StateSpace ssf = *cpCtrl->serialize(data);
    if(DEBUG) std::cout << "T0: " << t0 <<  "TF: " << tf << "DT: " << tf - ros::Time::now().toSec() << std::endl;
    cpCtrl->interpolater.compute_interpolation_params(ss0, ssf, t0, tf);
    cpCtrl->set_active();
}


///
/// \brief Interpolate::interpolate_cr
/// \param data
///
void Interpolate::interpolate_cr(_cr_data_type &data){
    if(DEBUG) std::cout << "Called: " << __FUNCTION__ << std::endl;
}

///
/// \brief Interpolate::interpolate_cv
/// \param data
///
void Interpolate::interpolate_cv(_cv_data_type &data){
    if(DEBUG) std::cout << "Called: " << __FUNCTION__ << std::endl;
}


///
/// \brief Interpolate::interpolate_cf
/// \param data
///
void Interpolate::interpolate_cf(_cf_data_type &data){
    if(DEBUG) std::cout << "Called: " << __FUNCTION__ << std::endl;
}


///
/// \brief Interpolate::interpolate_jp
/// \param data
///
void Interpolate::interpolate_jp(_jp_data_type &data){
    if(DEBUG) std::cout << "Called: " << __FUNCTION__ << std::endl;
    _jp_data_type robot_state;
    jpCtrl->robot_state_ifc->get_data(robot_state);
    double t0 = ros::Time::now().toSec();
    double tf = t0 + jpCtrl->compute_dt(t0);
    StateSpace ss0 = *jpCtrl->serialize(robot_state);
    StateSpace ssf = *jpCtrl->serialize(data);
    if(DEBUG) std::cout << "T0: " << t0 <<  "TF: " << tf << "DT: " << tf - ros::Time::now().toSec() << std::endl;
    jpCtrl->interpolater.compute_interpolation_params(ss0, ssf, t0, tf);
    jpCtrl->set_active();
}

///
/// \brief Interpolate::interpolate_jr
/// \param data
///
void Interpolate::interpolate_jr(_jr_data_type &data){
    if(DEBUG) std::cout << "Called: " << __FUNCTION__ << std::endl;
}


///
/// \brief Interpolate::interpolate_jv
/// \param data
///
void Interpolate::interpolate_jv(_jv_data_type &data){
    if(DEBUG) std::cout << "Called: " << __FUNCTION__ << std::endl;
}


///
/// \brief Interpolate::interpolate_jf
/// \param data
///
void Interpolate::interpolate_jf(_jf_data_type &data){
    if(DEBUG) std::cout << "Called: " << __FUNCTION__ << std::endl;
}


/////
/// \brief Move::Move
///
Move::Move(){
    method_map["move_cp"] = FcnHandleBasePtr(new FcnHandle<_cp_data_type>(&Move::move_cp, this));
    method_map["move_cr"] = FcnHandleBasePtr(new FcnHandle<_cr_data_type>(&Move::move_cr, this));
    method_map["move_jp"] = FcnHandleBasePtr(new FcnHandle<_jp_data_type>(&Move::move_jp, this));
    method_map["move_jr"] = FcnHandleBasePtr(new FcnHandle<_jr_data_type>(&Move::move_jr, this));
}

///
/// \brief Move::move_cp
/// \param data
///
void Move::move_cp(_cp_data_type &data){
    if(DEBUG) std::cout << "Called: " << __FUNCTION__ << std::endl;
}

///
/// \brief Move::move_cr
/// \param data
///
void Move::move_cr(_cr_data_type &data){
    if(DEBUG) std::cout << "Called: " << __FUNCTION__ << std::endl;
}

///
/// \brief Move::move_jp
/// \param data
///
void Move::move_jp(_jp_data_type &data){
    if(DEBUG) std::cout << "Called: " << __FUNCTION__ << std::endl;
}

///
/// \brief Move::move_jr
/// \param data
///
void Move::move_jr(_jr_data_type &data){
    if(DEBUG) std::cout << "Called: " << __FUNCTION__ << std::endl;
}

/////
/// \brief Servo::Servo
///
Servo::Servo(){
    method_map["servo_cp"] = FcnHandleBasePtr(new FcnHandle<_cp_data_type>(&Servo::servo_cp, this));
    method_map["servo_cr"] = FcnHandleBasePtr(new FcnHandle<_cr_data_type>(&Servo::servo_cr, this));
    method_map["servo_cv"] = FcnHandleBasePtr(new FcnHandle<_cv_data_type>(&Servo::servo_cv, this));
    method_map["servo_cf"] = FcnHandleBasePtr(new FcnHandle<_cf_data_type>(&Servo::servo_cf, this));
    method_map["servo_jp"] = FcnHandleBasePtr(new FcnHandle<_jp_data_type>(&Servo::servo_jp, this));
    method_map["servo_jr"] = FcnHandleBasePtr(new FcnHandle<_jr_data_type>(&Servo::servo_jr, this));
    method_map["servo_jv"] = FcnHandleBasePtr(new FcnHandle<_jv_data_type>(&Servo::servo_jv, this));
    method_map["servo_jf"] = FcnHandleBasePtr(new FcnHandle<_jf_data_type>(&Servo::servo_jf, this));
}

///
/// \brief Servo::servo_cp
/// \param data
///
void Servo::servo_cp(_cp_data_type &data){
    if(DEBUG) std::cout << "Called: " << __FUNCTION__ << std::endl;
    cpCtrl->robot_cmd_ifc->set_data(data);
}

///
/// \brief Servo::servo_cr
/// \param data
///
void Servo::servo_cr(_cr_data_type &data){
    if(DEBUG) std::cout << "Called: " << __FUNCTION__ << std::endl;
    crCtrl->robot_cmd_ifc->set_data(data);
}

///
/// \brief Servo::servo_cv
/// \param data
///
void Servo::servo_cv(_cv_data_type &data){
    if(DEBUG) std::cout << "Called: " << __FUNCTION__ << std::endl;
    cvCtrl->robot_cmd_ifc->set_data(data);
}

///
/// \brief Servo::servo_cf
/// \param data
///
void Servo::servo_cf(_cf_data_type &data){
    if(DEBUG) std::cout << "Called: " << __FUNCTION__ << std::endl;
    cfCtrl->robot_cmd_ifc->set_data(data);
}

///
/// \brief Servo::servo_jp
/// \param data
///
void Servo::servo_jp(_jp_data_type &data){
    if(DEBUG) std::cout << "Called: " << __FUNCTION__ << std::endl;
    jpCtrl->robot_cmd_ifc->set_data(data);
}

///
/// \brief Servo::servo_jr
/// \param data
///
void Servo::servo_jr(_jr_data_type &data){
    if(DEBUG) std::cout << "Called: " << __FUNCTION__ << std::endl;
    jrCtrl->robot_cmd_ifc->set_data(data);
}

///
/// \brief Servo::servo_jv
/// \param data
///
void Servo::servo_jv(_jv_data_type &data){
    if(DEBUG) std::cout << "Called: " << __FUNCTION__ << std::endl;
    jvCtrl->robot_cmd_ifc->set_data(data);
}

///
/// \brief Servo::servo_jf
/// \param data
///
void Servo::servo_jf(_jf_data_type &data){
    if(DEBUG) std::cout << "Called: " << __FUNCTION__ << std::endl;
    jfCtrl->robot_cmd_ifc->set_data(data);
}

///
/// \brief Controllers::Controllers
/// \param name_space
/// \param arm_name
///
Controllers::Controllers(){

}


///
/// \brief Controllers::get_method_by_name
/// \param method_name
/// \return
///
FcnHandleBasePtr Controllers::get_method_by_name(std::string method_name){
    std::vector<std::string> x = split_str(method_name, '_');
    std::string mode = x[0];
    char op_space = x[1][0];
    char controller = x[1][1];

    if (mode.compare("interpolate") == 0){
        return Interpolate::get_method_by_name(method_name);
    }
    if (mode.compare("move") == 0){
        return Move::get_method_by_name(method_name);
    }
    if (mode.compare("servo") == 0){
        return Servo::get_method_by_name(method_name);
    }
    else{
        throw "method name not understood";
    }
}

